core/src/style/pointStyleBuilder.cpp

```diff
@@ -329,39 +299,106 @@ bool PointStyleBuilder::getUVQuad(Parameters& _params, glm::vec4& _quad, Texture
 
     if (texture) {
         *_texture = texture;
+    }
 
-        if (auto& atlas = texture->spriteAtlas()) {
+    return true;
+}
 
-            SpriteNode spriteNode;
+bool PointStyleBuilder::evalSizeParam(const DrawRule& _rule, Parameters& _params, const Texture* _texture) const {
+    StyleParam::SizeValue size;
+    SpriteNode spriteNode;
+    glm::vec2 cssSize = {1.f, 1.f};
+    float aspectRatio = 1.f;
+    bool useTextureInfo = false;
+
+    if (_texture) {
+        cssSize = glm::vec2{_texture->getWidth(), _texture->getHeight()} * _texture->invDensity();
+        aspectRatio = _texture->getWidth() / _texture->getHeight();
+        // can use cssSize and aspect if "%" or "auto" size is used
+        useTextureInfo = true;
+
+        const auto &atlas = _texture->spriteAtlas();
+        if (atlas) {
             if (!atlas->getSpriteNode(_params.sprite, spriteNode) &&
                 !atlas->getSpriteNode(_params.spriteDefault, spriteNode)) {
                 return false;
             }
+            cssSize = spriteNode.m_size * _texture->invDensity();
+            aspectRatio = spriteNode.m_size.x / spriteNode.m_size.y;
+        } else {
+            // missing sprite atlas for texture but sprite specified in draw rule
+            return false;
+        }
+    } else {
+        // default css size of 16px
+        cssSize = glm::vec2(16.f);
+    }
+
+    auto sizeParam = _rule.findParameter(StyleParamKey::size);
+
+    if (sizeParam.stops) {
+        glm::vec2 lower = sizeParam.stops->evalSize(m_styleZoom, cssSize, aspectRatio, useTextureInfo);
+        // illegal size values were evaluated
+        if (std::isnan(lower.x)) { return false; }
+
+        glm::vec2 higher = sizeParam.stops->evalSize(m_styleZoom + 1, cssSize, aspectRatio, useTextureInfo);
+        _params.size = lower;
+        _params.extrudeScale = higher.x - lower.x;
 
-            if (std::isnan(_params.size.x)) {
-                // determine the css size of the sprite if size is not determined from style draw rule
-                _params.size = spriteNode.m_size * texture->invDensity();
+    } else if (_rule.get(StyleParamKey::size, size)) {
+        // illegal size values in draw rule
+        if ( (size[0].isAuto() && size[1].isAuto()) ||
+                (!useTextureInfo && (size[0].isPercentage() || size[0].isAuto() || size[1].isAuto()))) {
+            return false;
+        }
+
+        if (size[0].isPixel() && size[1].isPixel()) {
+            if (std::isnan(size[0].value) && std::isnan(size[1].value)) {
+                _params.size = cssSize;
+            } else if (size[0].value == 0.f || std::isnan(size[1].value)) {
+                _params.size = glm::vec2(size[0].value);
+            } else {
+                _params.size = glm::vec2(size[0].value, size[1].value);
             }
+        } else {
+            // _texture/spriteNode must have been set!
+            assert(useTextureInfo);
```

Comment:
Do we really want to assert here? This seems like a scene authoring error and that should not stop the program. Maybe I misunderstand this logic?

Reply:
Make sense, seems to be a job of a linter, but will change this to be a conditional check.